File: LeetCode/Medium/2597-the-number-of-beautiful-subsets/2597-the-number-of-beautiful-subsets.cpp
class Solution {
public:
    int n = 0;
    int ret = 0;
    int beautifulSubsets(vector<int>& nums, int k) {
        n = nums.size();
        vector<int> count(1001);

        dfs(0, count, nums, k);

        return ret;
    }
    
    void dfs(int idx, vector<int>& count, vector<int>& nums, int k){
        if (idx == n) return;

        int temp = nums[idx];

        if ((temp - k < 0 || count[temp - k] == 0) && (
            temp + k > 1000 || count[temp + k] == 0)){
            ++ret;
            ++count[temp];

            dfs(idx + 1, count, nums,k );
            count[temp]--;
            dfs(idx + 1, count, nums, k);
        } else dfs(idx + 1, count, nums, k);
    }
};

File: LeetCode/Medium/2597-the-number-of-beautiful-subsets/2597-the-number-of-beautiful-subsets.go
func beautifulSubsets(nums []int, k int) int {
    n := len(nums)
    ret := 0
    count := make([]int, 1001)
    
    dfs := func(idx int){}
    dfs = func(idx int){
        if idx == n { return }

        temp := nums[idx]
        
        if (temp - k < 0 || count[temp - k] == 0) && 
        (temp + k > 1000 || count[temp + k] == 0) {
            ret++
            count[nums[idx]]++

            dfs(idx + 1)
            count[temp]--
            dfs(idx + 1)
        } else { dfs(idx + 1) }
    }

    dfs(0)

    return ret
}